### PR TITLE
Version 2: rewrite with Angular

### PR DESCRIPTION
**Pull request by rob42**
Replacement for reverted pull request #46 #48

### CODE REVIEW

**Comment by tkurki** [comment]
To me this PR is _Rewrite as version 2_, how about renaming?

**Comment by tkurki** [comment]
Some deprecation warnings, not showstopper but maybe something to consider fixing:
```
source $ npm install
npm WARN deprecated openlayers@4.4.2: Please use the ol package instead
npm WARN deprecated uws@9.14.0: stop using this version
```

**Comment by tkurki** [comment]
Oh. openlayers is at 4.6.5 and has some security warnings from npm, ol is at 5.2.0. Consider upgrading?

https://github.com/openlayers/openlayers/releases/tag/v5.0.0 says _completely removing any dependency on the Closure Compiler_, which is one source of the high severity security issues.

**Comment by tkurki** [comment]
Ok, with the changes in #52 I can get the webapp to load via node server and render ok.

Steps:
- `npm run prepare`
- `npm pack`
- in node server:
- `npm install ../freeboard-sk/signalk-freeboard-sk-0.1.0.tgz`
- `bin/n2k-from-file`
- http://localhost:3000/@signalk/freeboard-sk/

But there is something funny still going on: the app is really sluggish and seems to get more so over time.

Here is a piece of code that you can use to monitor js running & blocking the timer from running: https://gist.github.com/tkurki/e635d2ceae45bedbace4933945916005

Paste that into your browser's console. With Freeboard my output looks like this
image

**Comment by tkurki** [comment]
The code in this branch publishes to `source/dist/freeboard` per [angular.json](https://github.com/SignalK/freeboard-sk/blob/dc5f32bf2aa57f53daf47f2a1ca1f6a2fd2a0e3c/source/angular.json#L16), but the assets are under root - and in git?

Having the assets under root does not work very well, because you are mixing persistent and transitory stuff and there is no easy way to clear/delete the artifacts.

I don't think it makes sense to include the artefacts in the git repo. It bloats the repo in size, the repo looks pretty ugly in Github and very few people benefit from it.

**Comment by tkurki** [comment]
Instead of a hardcoded url in dev mode I think it would make more sense to try to connect to the server where the app was loaded from. 

You can add an override for example from localstorage if you need to, but developing so that the app is loaded the way it works in the field saves you from a bunch of glitches.

**Comment by tkurki** [comment]
No sluggishness with a dev build.

**Comment by rob42** [comment]
Artemis server now uses the following chart format, and works nicely.
```
{
  "urn:mrn:signalk:uuid:80f1c606-4454-4d13-8ffc-2af0823cf931": {
    "identifier": "NZ61402",
    "format": "png",
    "description": "Port Motueka",
    "scale": 20000,
    "type": "tilelayer",
    "minzoom": 13,
    "maxzoom": 16,
    "name": "Port Motueka",
    "bounds": [
      173.01664692588037,
      -41.15666787764086,
      173.05998577188987,
      -41.13166913951548
    ],
    "tilemapUrl": "/./mapcache/NZ61402/{z}/{x}/{-y}.png",
    "timestamp": "2018-09-26T22:15:08.191Z",
    "$source": "unknown"
  },
```

**Comment by panaaj** [comment]
> Oh. openlayers is at 4.6.5 and has some security warnings from npm, ol is at 5.2.0. Consider upgrading?
> 
> https://github.com/openlayers/openlayers/releases/tag/v5.0.0 says _completely removing any dependency on the Closure Compiler_, which is one source of the high severity security issues.

The `ngx-openlayers` angular library used is currently being ported to Open Layers 5 so that will allow the move away from the current v4 issues.
Will update the library as soon as it's available.

**Comment by panaaj** [comment]
Have pushed commit [#f373a8bd9aeee87b86d6a7aea732ba042a3bda65](https://github.com/SignalK/freeboard-sk/pull/51/commits/f373a8bd9aeee87b86d6a7aea732ba042a3bda65) to address AIS target rendering performance.

Update also addresses:
- AIS target heading display - now reflect COG where previously they all pointed east.
- Displays labels only at a zoom level of 10 or greater.

File: source/src/app/lib/components/feature-ais.component.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            );
+            let f=layer.getFeatureById('ais-'+ e[0]);
+            if(f) {
+                f.setGeometry( new geom.Point(tc) );
+                f.setStyle( 
+                    new style.Style({
+                        image: new style.Icon({
+                            src: this.icon,
+                            rotateWithView: true,
+                            rotation: e[1].cogTrue
```
Comment (joabakk) [inline source/src/app/lib/components/feature-ais.component.ts]: COG makes sense when vessels are at speed, but once anchored or moored, heading would be more stable and correct.

**Comment by joabakk** [comment]
> 
> 
> How about updating the UI to display the following:
> 
>     * magenta icon = active AIS target (update received in last 3 mins)
> 
>     * brown icon - inactive AIS target (no update received in last 6 mins - missed at least two update cycles)
> 
>     * delete icon after three missed cycles i.e. 9 mins
>       image
> 
> 
> Thoughts?

Would there be any way for the user to configure these limits?

**Comment by tkurki** [comment]
Let’s not get distracted - configurability can be in the backlog, we need to publish the new version first.

**Comment by tkurki** [comment]
I tried with bin/n2k-from-file and things work ever so smoothly now!

I'd be happy to publish this as it stands, except for npm install / audit complains about `found 242 vulnerabilities (227 low, 10 moderate, 5 high)`. I can still publish this, as most of the complains are not relevant for a webapp, and I'd hate to have this stall. But at the same time I'd love to see at least the low hanging npm audit fruits picked!

**Comment by tkurki** [comment]
Then there's the conflicts to take care of.

**Comment by panaaj** [comment]
> I'd be happy to publish this as it stands, except for npm install / audit complains about `found 242 vulnerabilities (227 low, 10 moderate, 5 high)`. 

With the current package.json and with a clean install of packages using @angular/cli 6.2.3 `npm audit` gives me the following result: 
`found 23 vulnerabilities (12 low, 6 moderate, 5 high) in 38827 scanned packages`

The `5 high` are within the `ngx-openlayers` components which currently require OpenLayers 4.x.x.
These are currently in the process of being ported to OpenLayers 5 and can be upgraded when available.

**Comment by panaaj** [comment]
> Then there's the conflicts to take care of.

I might leave this conflict resolution to more experienced GitHub operators... the files listed should be straight forward enough to resolve:
- .gitignore: use content from `panaaj-angular` branch
- .npmignore; use content from `panaaj-angular` branch
- CHANGELOG.md: new file
- README.md: use content from `panaaj-angular` branch
- package-lock.json: remove
- package.json: use content from `panaaj-angular` branch

**Comment by tkurki** [comment]
The vulnerability stats I gave were from current package.json with a clean install with `npm install` - what does _clean install of packages using @angular/cli 6.2.3_  mean? Is its use documented somewhere?

See https://gist.github.com/tkurki/e0aaf5dc1e5c11aeb853fbf6f674325c

**Comment by panaaj** [comment]
The recommended method is:
1. Delete `node_modules` folder
2. Run `npm i —save-dev @angular/cli@latest`
3. Run `npm I`

@angular/cli github site.

**Comment by tkurki** [comment]
Why should the cli be upgraded always, independent of other deps in package.json? Won’t this break when cli gets a major update?

**Comment by panaaj** [comment]
It doesn’t always have to be updated.. it’s just in this instance it was updated during the development process due to some cli dependencies having vulnerabilities.... so the sure fire way of ensuring these files are not used it was recommended the node_nodules folder be deleted, and latest cli installed from scratch.

**Comment by tkurki** [comment]
We need to have a version that works out of git as one might expect.

I might figure out that one is expected to run `npm install` in `source` director, so the vulnerabilities from that installation count. Not from running something else, like _Run npm i —save-dev @angular/cli@latest_.

Like, why urge me to run `npm i —save-dev @angular/cli@latest` instead of commiting the changes from that to package.json, so that I could just pull, nuke node_modules, install and `npm pack` to produce a build that is (mostly) repeatable and with just the minimum necessary vulnerabilities?

**Comment by sbender9** [comment]
I'm unable to build this:

```
$ npm i

> @signalk/freeboard-sk@1.0.0 prepare /usr/local/src/freeboard-sk.master
> cd source && npx ng build --prod

Could not find module "@angular-devkit/build-angular" from "/usr/local/src/freeboard-sk.master/source".
Error: Could not find module "@angular-devkit/build-angular" from "/usr/local/src/freeboard-sk.master/source".
```

**Comment by panaaj** [comment]
Just so we are all on the same page:

To build the Angular project in `source` you need the `@angular/cli` which is th `ng` in `ng build --prod`
You would normally install this globally so you can use it to generate new projects, etc.. `npm i -g @angular/cli@latest`.

So to build the project the process would be:
1. clone from GitHub
2. `cd source`
3. `npm i` to install Angular modules, the CLI and other dependencies
The result of which should be a successful install of `node_modules` and the result for me is:
```added 1281 packages from 1348 contributors and audited 38827 packages in 134.83s
  found 23 vulnerabilities (12 low, 6 moderate, 5 high)
  run `npm audit fix` to fix them, or `npm audit` for details
```
4. run `ng build --prod` to build the project.
The built application files will be placed in `../public` folder which can then be packaged for `npm`.

@sbender9 the error you received is because npm pack is trying to run `ng build --prod` and the `@angular/cli` is not installed.... need to `npm i` in `source` first.

**Comment by sbender9** [comment]
OK, got it to build.It's not working when being loaded via https:

```[Warning] [blocked] The page at https://localhost:3443/@signalk/freeboard-sk/ was not allowed to run insecure content from ws://localhost:3443/signalk/v1/stream. (bundle.js, line 3779)
```

I have not been able to figure out how to fix it...

**Comment by tkurki** [comment]
What version of Node do you have? With v8.12.0 and a fresh clone from Github npm install gives me `242 vulnerabilities (227 low, 10 moderate, 5 high)`.

I just published this as @signalk/freeboard-sk@1.0.0-0, a pre release version. @panaaj would you like to promote that a bit on Slack? You do need to install it manually, in ~/.signalk/ `npm install @signalk/freeboard-sk@1.0.0-0`.

Is there a reason to keep things under `source` as opposed to having all the files that are there directly under root and dependencies as dev dependencies in root package.json? .npmignore should be naturally updated. 

That way development would be simpler and like in any other project. 

Hey, we are (or more like Adrian is) close to finishing this!

**Comment by panaaj** [comment]
node v8.11.1
npm 6.4.1

**Comment by tkurki** [comment]
Readme needs an update, at least the pics.

**Comment by tkurki** [comment]
But there is nothing but the angular files in this project - what would be the potential issues?

**Comment by panaaj** [comment]
> But there is nothing but the angular files in this project - what would be the potential issues?

If you just want the Angular project and no other files then no problem... I thought we needed the NPM package files at the root for Artemis?

**Comment by panaaj** [comment]
README.md images updated

**Comment by joabakk** [comment]
Finally got around to taking it for a spin. I think it looks very good, great work! The only thing I am not overly enthusiastic about is that it looks for self, and if none is found, assumes that there is a vessel at equator, 0 deg E/W. Again this may not be an issue for most use.

**Comment by joabakk** [comment]
Sorry to be slow, but from which floder is `ng build` run? I follower the explanation to here, but I get `-bash: ng: command not found`

**Comment by tkurki** [comment]
This is sort of the point: a potential contributor should be able to do things the way one expects, which to me is

- `npm install`
- hmmm how do I build this thing, I'll  try `npm run build` and/or `npm run watch`
- (well those didn't work, let's run `npm run` to see what there is

**Comment by joabakk** [comment]
I've noticed that if a browser tab with the new version has been open for a while, the vessels disappear and I have to refresh to get them back. This is after I have manually changed the timeouts from 6/9 min to 1/24h and rebuilt, but otherwise no changes

**Comment by tkurki** [comment]
Guys, are we done with the rewrite - time to merge? 

Merge by overwriting things in master by latest in this branch?

**Comment by rob42** [comment]
I vote merge - up to @panaaj now :-)

**Comment by panaaj** [comment]
Merge...and yes overwrite everything

**Pull request merged**